octodns/record/svcb.py

```diff
@@ -0,0 +1,251 @@
+    try:
+        port = int(svcparamvalue)
+        if 0 < port > 65535:
+            reasons.append(f'port {port} is not a valid number')
+    except ValueError:
+        reasons.append('port is not a number')
+    return reasons
+
+
+def validate_svcparam_alpn(svcparamvalue):
+    reasons = []
+    alpns = svcparamvalue.split(',')
+    for alpn in alpns:
+        reasons += _ChunkedValue.validate(alpn, 'SVCB')
+    return reasons
+
+
+def validate_svcparam_iphint(ip_version, svcparamvalue):
+    reasons = []
+    addresses = svcparamvalue.split(',')
+    for address in addresses:
+        try:
+            if ip_version == 4:
+                IPv4Address(address)
+            if ip_version == 6:
+                IPv6Address(address)
+        except AddressValueError:
+            reasons.append(
+                f'ip{ip_version}hint "{address}" is not a valid IPv{ip_version} address'
+            )
+    return reasons
+
+
+def validate_svcparam_ip4hint(svcparamvalue):
+    return validate_svcparam_iphint(4, svcparamvalue)
+
+
+def validate_svcparam_ip6hint(svcparamvalue):
+    return validate_svcparam_iphint(6, svcparamvalue)
+
+
+def validate_svcparam_mandatory(svcparamvalue):
+    reasons = []
+    mandatories = svcparamvalue.split(',')
+    for mandatory in mandatories:
+        if (
+            mandatory not in SUPPORTED_PARAMS.keys()
+            and not mandatory.startswith('key')
+        ):
+            reasons.append(f'unsupported SvcParam "{mandatory}" in mandatory')
+        if mandatory.startswith('key'):
+            reasons += validate_svckey_number(mandatory)
+    return reasons
+
+
+def validate_svcparam_ech(svcparamvalue):
+    try:
+        b64decode(svcparamvalue, validate=True)
+    except binascii_error:
+        return ['ech SvcParam is invalid Base64']
+
+
+def validate_svckey_number(paramkey):
+    try:
+        paramkeynum = int(paramkey[3:])
+        if 7 < paramkeynum > 65535:
+            return [f'SvcParam key "{paramkey}" has wrong key number']
+    except ValueError:
+        return [f'SvcParam key "{paramkey}" has wrong format']
+    return []
+
+
+SUPPORTED_PARAMS = {
+    'no-default-alpn': {'has_arg': False},
+    'alpn': {'validate': validate_svcparam_alpn},
+    'port': {'validate': validate_svcparam_port},
+    'ipv4hint': {'validate': validate_svcparam_ip4hint},
+    'ipv6hint': {'validate': validate_svcparam_ip6hint},
+    'mandatory': {'validate': validate_svcparam_mandatory},
+    'ech': {'validate': validate_svcparam_ech},
+}
+
+
+class SvcbValue(EqualityTupleMixin, dict):
+
+    @classmethod
+    def parse_rdata_text(cls, value):
+        try:
+            # XXX: these are called SvcPriority, TargetName, and SvcParams in RFC 9460 section 2.
+            #       Should we mirror these names, or are priority, target and params good enough?
+            # XXX: Should we split params into the specific ParamKeys and ParamValues?
+            (priority, target, *params) = value.split(' ')
```

Comment:
This one is interesting. Generally octoDNS has tended to leave "compound" params like this alone and just use the strings verbatim, but for the most part those fields didn't explicitly say they're a list of key & value pairs and here that is the case and there's even further limiting things to only appearing once and laying out the details of the k & v.

https://datatracker.ietf.org/doc/html/rfc9460#presentation

> Each SvcParamKey SHALL appear at most once in the SvcParams

https://datatracker.ietf.org/doc/html/rfc9460#keys

I'd lean toward parsing them and having them as a dict in octoDNS's model of things, and thus in the YAML. I think that'd mean this bit stays as-is and something down below near the `unquote(target)` would convert the list of strings into a dict (with `None` for the value of anything is valueless) unless I've missunderstood the details of the params. 

I guess there's a valid question of whether we further parse things like ipv4/6hint and turn the `,` separated IPs into lists.

Reply:
There them is the question of having keys like this:

```
svcparams:
  alpn: h2,h3
  ipv4hint: 192.0.2.0
```

or with the actual titles:

```
svcparams:
  - svcparamkey: alpn
    svcparamvalue: h2,h3
  - svcparamkey: ipv4hint
    svcparamvalue: 192.0.2.1
```

I'm leaning towards the first  to make it easy on the users.

Reply:
> I guess there's a valid question of whether we further parse things like ipv4/6hint and turn the , separated IPs into lists.

The real question is, what does octodns gain from that?

Reply:
alright, I'll update that. I wonder how big the mental strain on users will be though. But with having the yaml under version control, changes are very easy to see.

```yaml
svcparams:
  alpn:
    - h2
    - h3
  ipv4hint:
    - 192.0.2.0
```

Or would you want accept this as well:

```yaml
svcparams:
  alpn:
    - h2
    - h3
  ipv4hint: 192.0.2.0
```

Reply:
> But with having the yaml under version control, changes are very easy to see.

Yeah, that's a main argument for breaking things out.



> Or would you want accept this as well:

No strong opinion. If it's easy to support the single value option is probably nice to have and something people will accidentally do from time to time.

Reply:
it might be nice to have, but I realize that would mean even more code to prevent users from not just stuffing a comma-separated string in the yaml. Explicitly requiring a list is cleaner.